firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift

```diff
@@ -1725,6 +1734,68 @@ class BrowserViewController: UIViewController,
         } else if state.showQRcodeReader {
             navigationHandler?.showQRCode(delegate: self)
         }
+
+        handleNavigationActions(for: state)
+    }
+
+    private func dispatchBackForwardToolbarAction(_ isEnabled: Bool?, _ windowUUID: UUID, _ actionType: ToolbarActionType) {
+        let action = ToolbarAction(isEnabled: isEnabled, windowUUID: windowUUID, actionType: actionType)
+
+        switch actionType {
+        case .backButtonStatus,
+             .forwardButtonStatus:
+            store.dispatch(action)
+        default: break
+        }
+    }
+
+    private func getNavigationAction(for state: BrowserViewControllerState,
+                                     actionType: ToolbarState.ActionState.ActionType) -> Bool {
+        let navigationToolbarAction = actionType == .back ?
+            state.toolbarState.navigationToolbar.actions.first(where: { $0.actionType == actionType })?.navigateBack :
+            state.toolbarState.navigationToolbar.actions.first(where: { $0.actionType == actionType })?.navigateForward
+
+        let addressToolbarAction = actionType == .back ?
+            state.toolbarState
+                 .addressToolbar
+                 .navigationActions
+                 .first(where: { $0.actionType == actionType })?.navigateBack :
+
+            state.toolbarState
+                 .addressToolbar
+                 .navigationActions
+                 .first(where: { $0.actionType == actionType })?.navigateForward
+
+        return navigationToolbarAction ?? addressToolbarAction ?? false
```

Comment:
can you explain what this method is doing, whats the distinction between navigation and address? Wondering if we can simplify...

Reply:
We have two `BrowserAddressToolbar` instances unfortunately, so we have to determine from which one the actions are coming.